csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs

```diff
@@ -941,11 +950,11 @@ private void InitWithSessionHandle(IntPtr session, SessionOptions options)
                     var inputMeta = GetInputMetadata(i);
                     var iname = GetInputName(i, out byte[] utf8);
                     _inputNames.Add(iname);
-                    inputMeta.ZeroTerminatedUtf8Name = utf8;
+                    var namePin = new Memory<byte>(utf8).Pin();
+                    _namesMemoryHandles.Add(namePin);
```

Comment:
This creates a `GCHandle` per entry and while allocating the handle isn't terribly expensive (most of the time), it can have a large negative downstream impact on the GC as it limits the ability for the GC to compat memory and causes it to need to work around every pin.

In .NET Core, the POH (pinned object heap) was introduced as a way to have long-term pinned data without negatively impacting the normal heap.

It would likely be better to allocate such data in native and track a pointer. Disposal would then call `Free` on the allocations instead. This will avoid the additional GC pressure and GC pessimizations caused by long term pinning of many objects.

The same goes for all the other `Memory<T>.Pin()` occurences

Reply:
Good to know about native memory, however, it is available since .NET 7